### ISSUE TITLE
DB file handler not closing: running out of max open file descriptors.

### ISSUE THREAD

**Issue by davidnewhall** [crawshaw/sqlite#67]
Hello,

I've tried version 0.1.2 as well as master. I feel like my code is pretty simple. I call `sqlite.OpenConn()` in read-only mode, run one `SELECT`, then call `conn.Close()`. 

Every invocation of this workflow opens more file descriptors for the DB and the wal file. Eventually my environment hits the max open file limit and the app stops working. I'm not sure what else to try here and any suggestions are appreciated.

This is a link to the code that is causing this problem:
https://github.com/golift/imessage/blob/master/imessage.go#L136-L157
```
motifini  75096        username   42u      REG               1,15     3184792   44580185 /Users/username/Library/Messages/chat.db-wal
motifini  75096        username   43r      REG               1,15     4608000   44580182 /Users/username/Library/Messages/chat.db
motifini  75096        username   44u      REG               1,15     3184792   44580185 /Users/username/Library/Messages/chat.db-wal
motifini  75096        username   45r      REG               1,15     4608000   44580182 /Users/username/Library/Messages/chat.db
motifini  75096        username   46u      REG               1,15     3184792   44580185 /Users/username/Library/Messages/chat.db-wal
motifini  75096        username   47r      REG               1,15     4608000   44580182 /Users/username/Library/Messages/chat.db
motifini  75096        username   48u      REG               1,15     3184792   44580185 /Users/username/Library/Messages/chat.db-wal
```
... and it goes on and on.

Thank you,
-david

**Comment by davidnewhall** [comment]
I've also tried adding `query.Finalize()` but that causes a nil pointer dereference. `query.Reset()` doesn't cause an error, but it also doesn't fix the problem. To make sure this is the sqlite library I also disabled the `fsnotify` code in that linked library. I cannot for the life of me figure out the trick to getting this library to close the db connection, and the open files.

**Comment by AdamSLevy** [comment]
I'm not the maintainer of this repo but I have been using it a lot lately and making some improvements on it.

I can assure you that the library properly closes database connections. 

It is unlikely that someone else is going to debug your code for you. If you truly believe there is an issue with this library, please create the simplest possible example that demonstrates the issue.

Just to clarify, if you use Prep or Prepare for your queries you should *not* call Finalize() on the Stmt, as they are cached and reused. If you use PrepareTransient, you must call Finalize() on the Stmt, when you are done using it.

**Comment by davidnewhall** [comment]
Run this on any mac.
```
package main

import (
	"fmt"
	"os"
	"time"

	"crawshaw.io/sqlite"
)

func main() {
	ticker := time.NewTicker(10 * time.Millisecond)
	var i int
	for range ticker.C {
		i++
		pollDB(i, os.Getenv("HOME")+"/Library/Messages/chat.db")
	}
}

func pollDB(count int, path string) {
	db, err := sqlite.OpenConn(path, sqlite.SQLITE_OPEN_READONLY)
	if err != nil {
		panic(err)
	}
	defer db.Close()

	query := db.Prep(`SELECT MAX(rowid) AS id FROM message`)
	defer query.Reset()
	if hasrow, err := query.Step(); err != nil {
		panic(err)
	} else if !hasrow {
		panic("no message rows found")
	}

	fmt.Println(count, "current iMessage rowID:", query.GetInt64("id"))
}

```
In another terminal window run `lsof | grep chat.db` a few times. The database file opens repeatedly until things crash. Sending (or receiving) messages with Messages.app while this is running will increment the rowID.

I get to 2430 before it panics. Every time.
```
2430 current iMessage rowID: 23148
panic: sqlite.OpenConn: SQLITE_CANTOPEN (/Users/newhalld/Library/Messages/chat.db)

goroutine 1 [running]:
main.pollDB(0x97f, 0xc0002fd8c0, 0x28)
	/test/main.go:23 +0x2dd
main.main()
	/test/main.go:16 +0xdd
```

Thanks for the help!

**Comment by davidnewhall** [comment]
I'm totally willing to accept that I'm doing something wrong here, and really appreciate someone pointing that out (debugging my code). I'm at a loss, and the above example points out the problem. This could be a macOS issue....

**Comment by AdamSLevy** [comment]
Thanks. I may test this later this week.

Off top though I can say that I agree that this code should be properly
opening, using, and closing database connections. It is hard to say what or
where the underlying issue is. However this strikes me as a very odd
usecase. Most programs open one database connection, or a pool of
connections, and then keep them open and in use until the end of the
program. Opening and closing thousand of connections for the same database
is non standard at least. Not reusing the database connection also stops
your program from taking advantage of the Stmt caching that Prep and
Prepare do. You'd be better off using PrepareTransient in such case because
you are only preparing a query once.

So for these reasons you might consider redesigning this aspect of your
application. Consider using a pool of readonly connections for concurrent
reading of the same database. See sqlite.Pool


Another side note, that is technically unrelated to your issue, is that
when using Stmt from Prep or Prepare, you don't need to reset them. You
only need to reset the statement if you want to hang onto the Stmt yourself
and restart the query from the first step. But in this example it is
superfluous. The next call to Prep on that same database for the same query
will use the cached Stmt and reset it before returning it to you for use.

Hope this helps get you going for now.

**Comment by davidnewhall** [comment]
Thanks for your response Adam, much appreciated! The snippet I provided here is a working example that reproduces the panic. I tested it on two macs before posting here. I provided it because you requested it, rightfully so. 

This is not my use case. :) It's just a snippet of code I wrote in a couple minutes to repo the problem I run into after my application runs for a few days. 

I understand I do not need to call Reset or Finalize, these were tried just to rule out any other issues. I'm not "up" on sqlite terms, unfortunately, so I was taking stabs in the dark. I'm very stumped. 

This library has served me well, and I'd love to continue using it. I'm still trying to debug this, trust me. I don't really expect anyone to debug my code for me, but it looks like a bug, so I reported it. Thank you for your patience and understanding.

**Comment by AdamSLevy** [comment]
I need to slightly amend what I said before about not needing to call Reset. It is necessary on Stmts that haven't completed (i.e. returned Step with err != nil or hasRow == false) if you are returning the Stmt's associated Conn to a Pool.

**Comment by davidnewhall** [comment]
I've modified my library to use PrepareTransient/Finalize instead of Prep(or Prepare)/Reset. I'm no longer having the issue I originally reported. It works fine on the latest release as well as master.

If the test code I provided above is supposed to close file handles, then I'm still going to suggest the sqlite library has a bug that's leaving them open. It is no longer affecting me though.

Thank you for all the information!
